[#106191288] Add logsearch for Cloud Foundry components and Apps

actionjack opened the pull request:
[Aggregate system logs ](https://www.pivotaltracker.com/story/show/106191288)
## What

To support the trial we need to have aggregated logs for the system to identify where issues are. We should aim for the simplest possible solution that would allow us to see the logs for all CF components and nodes in a single location to allow for easy debugging. Each log entry should identify which component and node generated the entry.

**Acceptance Criteria**
**Given** that there is an issue that I am attempting to debug
**When** look at the aggregated CF system logs
**Then** I can see all the log entries generated by all CF components
**And** each entry clearly indicates the component that generated the entry
**And** each entry clearly indicates that the node that generated the entry
## How this PR should  be reviewed

This PR should be review in the following context:
- I want to use [logsearch-for-cloudfoundry](https://github.com/logsearch/logsearch-for-cloudfoundry) logsearch addon to create a custom [Elasticsearch-Logstash-Kibana](https://www.elastic.co/products) stack deployment
- I want to use [Terraform](https://terraform.io/) to create the right network resources so I can span it across 2 Availability Zones and route all it's traffic through the bastion box.
- The current `logsearch-deployment.yml` file fails because the collector template has now moved to be part of the cf release I want to fix this.
- The current `logsearch-jobs.yml` file has lots of unnecessary white spaces that can be removed to make the code cleaner and cross IDE friendly.
- The current `logsearch-jobs.yml` deploys too many VMs, I want to have the option to deploy all the services in one VM, by adding a `logsearch-minimal-jobs.yml`
- I want to create a logsearch deployment that works for `AWS`
- I want to upload `all the things` and have a script to generate a logsearch deployment yaml file.
- I want to integrate the deployment with Makefile tasks.
- I want all the components from CF to send logs to logsearch, by setting the right configuration on the metron agents.
- I want a consistent deployment name as  `logsearch-<envname>`
- I want to create a logsearch deployment that works for `GCE`, for which I need to use DNS to point the services instead of static IPs.
- I want to be able to deploy `kibana-for-cloudfoundry` using the errand from `logsearch-for-cloudfoundry`, for what I need set the IP of elasticsearch server in the manifest after deploying it.
- I want `collector` to connect to nats on `AWS` and `GCE`, using DNS on GCE instead of static IPs.
## How to Test this PR
- Deploy an environment in both AWS and GCE (it is OK if you already have an environment):
  - AWS `make aws DEPLOY_ENV=$MYENVIRONMENT`
  - GCE `make aws DEPLOY_ENV=$MYENVIRONMENT`
- Connect using ssh to the bastion box (`make ssh-aws` or `make ssh-gce`)
- To deploy, run the following on the bastion after deploying CF:

```
./scripts/deploy_logsearch.sh aws
```

  or 

```
./scripts/deploy_logsearch.sh gce
```
- Access the board in: https://logs.$MYENVIRONMENT.cf.paas.alphagov.co.uk/ or https://logs.$MYENVIRONMENT.cf2.paas.alphagov.co.uk/ 
## Who should review this PR
- A handsome hedgehog, if a good looking spiny mammal is unavailable anyone in the core team can step up to the plate except @keymon or @actionjack

keymon committed:
Copy templates from logsearch-for-cloudfoundry

keymon committed:
Create a new AWS subnet for log search services

Add also the right outputs and variables.

keymon committed:
Use collector template from cf release

The cf release includes the collector template,
not the `logsearch-for-cloudfoundry` one

keymon committed:
Cleaning formatting

keymon committed:
Add AWS infra and stubs for logsearch

Hardcoding some stuff like nats and so on, just to make it work

keymon committed:
Add script to generate the logsearch manifest

Make will upload it, but currently it does not deploy it.

To deploy, run this on the bastion after deploying CF:

```
./generate_logsearch_manifest.sh aws > logsearch_manifest.yml
bosh deployment logsearch_manifest.yml
bosh deploy
bosh run errand push-kibana
```

Access the board in:

https://logs.hector051015.cf.paas.alphagov.co.uk/

actionjack committed:
Add Cloud Foundry releases for logsearch

Add Cloud Foundry releases for:

* `logsearch` and
* `logsearch-for-cloudfoundry`

keymon committed:
Add a minimal logsearch with everything in one VMs

actionjack committed:
Script to compile and deploy logsearch manifest

Create script to compile and deploy logsearch release manifest.

actionjack committed:
Add makefile target to create logsearch deployment

actionjack committed:
Set syslog daemon target to logsearch deployment

keymon committed:
Include environment name in logsearch deploy name

In a similar way to `cloudfoundry-<envname>` we call the log search
deployment: `logsearch-<envname>`

keymon committed:
Allow parametrise IPs of log search jobs

This way we can optionally not use static IPs, but DNS names on GCE.

keymon committed:
Add logsearch infrastructure template for GCE

Define the required machine pools and networks. Note that
the network must be tagged as `cf` to allow route via the
nat box.

keymon committed:
Add additional outputs for logsearch on GCE

To configure log search on GCE we need to know the root CF domain
and the log search network name (in this case, we use the same as
the rest of deployments).

keymon committed:
Allow generate logsearch manifest for GCE

keymon committed:
Use logsearch-minimal-jobs template

This way we run all the log search required services in one unique VM.

keymon committed:
Enable syslog forwarding for CF on GCE

As static IPs are not supported on GCE, we must use the powerDNS Bosh DNS name.

keymon committed:
Fix typos and missconfig in logsearch minimal conf

keymon committed:
Stub elasticsearch IP from bosh output in GCE

Currently the errand that deploys kibana-for-cloudfoundry tries to setup a security-group on CF based to allow access to elasticsearch based on the IP defined in the manifest. There is no way to 
specify an IP range and DNS aliases are not valid in this case.

On GCE This is an issue, as GCE CPU does not support static IPs, 
but the VMs get IPs dynamically allocated.

The solution requires either do a pull request or deploy kibana out 
of the errand, but as a **sort term** solution we will workaround 
this by:

 1. Deploy logsearch (Elasticsearch+Lostash as usual)
 2. Query bosh to get the allocated IP.
 3. Generate a spiff stub with the IP for eleasticsearch.
 4. Run the errand with the new config.

keymon committed:
Parametrise the nats addresses for collector job

In the logsearch manifest, we need to define the nats addresses to
for the collector service to connect, which monitors the CF components.

In AWS nats have static IPs, but on GCE we use powerDNS bosh aliases.

dcarley commented on manifests/templates/logsearch/logsearch-deployment.yml:
```diff
@@ -0,0 +1,78 @@
+    # Default vcap & root password on deployed VMs (ie c1oudc0w)
+    # Generated using mkpasswd -m sha-512
+    bosh:
+      password: (( merge || "$6$4gDD3aV0rdqlrKC$2axHCxGKIObs6tAmMTqYCspcdvQXh3JJcvWOY2WGb4SrdXtnCyNaWlrf3WEqvYR2MYizEGp3kMmbpwBC6jsHt0" ))
```
Should this be in pass? Are we overriding it? If so, should we omit the default?

keymon replied:
Good catch, actually we missed overriding the passwords properly. I will fix that.

keymon replied:
Fixed, I override that. 

Do you want to omit the default?

dcarley commented on manifests/templates/logsearch/logsearch-infrastructure-warden.yml:
```diff
@@ -0,0 +1,463 @@
+    name: random
+
+networks: (( merge || networks_default ))
+networks_default:
```
I'm confused; what are all of these networks listed below?

keymon replied:
This is indeed really confusing. This is actually a workaround of a issue in the warden bosh CPI, which forces you to define all those networks.

This is only required for warden, and we don't really use it, unless you deploy this on bosh-lite.

keymon committed:
Override passwords for vcap user and kibana oauth2

mtekel commented on manifests/templates/logsearch/logsearch-infrastructure-aws.yml:
```diff
@@ -0,0 +1,75 @@
+    availability_zone: (( meta.zones.z1 ))
+- name: high_cpu
+  cloud_properties:
+    instance_type: r3.large
```
How about c4.large? r3 instances are high memory, r3.large has 15G ram. C4 are compute, better corresponding to the GCE highcpu instances...

keymon replied:
It makes sense I will change it. Right we only use the high_memory pool, as we have only one instance using the `logsearch-minimal-jobs.yml` template.

keymon replied:
done

keymon committed:
Logsearch high_cpu instance type is c4.large

This makes more sense for high computational VMs and will be aligned
with the machine type in GCE.

mtekel commented:
There's one small nuisance this creates: When you have 2 deployments, bosh will have the last deployment selected - which is logsearch. You will not be able to do BOSH cli operations on the jobs from the main deployment until you select it. For that you need a deployment file. Luckily if you are on bastion the file is there (cf-manifest.yml). Otherwise you can still download manifest using bosh.

mtekel commented:
Works on GCE and AWS as well, merging.
